Engine/ac/string.cpp

```diff
@@ -49,6 +49,67 @@ int String_IsNullOrEmpty(const char *thisString)
     return 0;
 }
 
+void calculate_joint_length(const size_t count, const uint32_t *&handles, int sep_len, int sep_ulen, int &total_len, int &total_ulen)
+{
+    total_len = 0;
+    total_ulen = 0;
+    for (uint32_t i = 0; i < count; ++i)
+    {
+        auto address = ccGetObjectAddressFromHandle(handles[i]);
+        if (address != nullptr) {
+            auto const& sh = ScriptString::GetHeader(address);
+            total_len += sh.Length;
+            total_ulen += sh.ULength;
+        }
+        if (i != count - 1) {
+            total_len += sep_len;
+            total_ulen += sep_ulen;
+        }
+    }
+}
+
+const char * String_Join(const char *separator, void* arrobj)
+{
+    if(arrobj == nullptr)
+        return nullptr;
+
+    auto const& header = CCDynamicArray::GetHeader(arrobj);
+    const size_t count = header.ElemCount;
+    if(count == 0)
+        return CreateNewScriptString("");
+
+    if(separator == nullptr)
+        separator = "";
+
+    const uint32_t *handles;
+    handles= static_cast<const uint32_t*>(arrobj);
+
+    int sep_len, sep_ulen;
+    ustrlen2(separator, &sep_len, &sep_ulen);
+
+    int total_len, total_ulen;
+    calculate_joint_length(count, handles, sep_len, sep_ulen, total_len, total_ulen);
+    auto buf = ScriptString::CreateBuffer(total_len, total_ulen);
+
+    // concatenate the array elements with a separator between elements: el1+sep+el2+sep...elN
+    char* p = buf.Get();
+    for (uint32_t i = 0; i < count; ++i)
+    {
+        auto address = (const char *) ccGetObjectAddressFromHandle(handles[i]);
+        if (address != nullptr) {
+            auto const& sh = ScriptString::GetHeader(address);
+            memcpy(p, address, sh.Length);
+            p += sh.Length;
```

Comment:
if you use std::copy, it returns iterator after copied fragment, letting you write slightly less lines:
```p = std::copy(address, address +sh.Length, p);```

Reply:
good, I didn't knew this, changed but didn't push force - will squash commits things later